test/OtokenFactory.test.ts

```diff
@@ -0,0 +1,221 @@
+        ethAddress,
+        usdc.address,
+        usdc.address,
+        strikePrice,
+        expiry,
+        isPut,
+        {from: accounts[0]},
+      )
+      expectEvent(txResponse, 'OtokenCreated', {
+        creator: accounts[0],
+        underlying: ethAddress,
+        strike: usdc.address,
+        collateral: usdc.address,
+        strikePrice: strikePrice.toString(),
+        expiry: expiry.toString(),
+        isPut: isPut,
+        tokenAddress: targetAddress,
+      })
+      otoken = await MockOtoken.at(targetAddress)
+    })
+
+    it('Should have correct paramter', async () => {
+      assert.equal(await otoken.underlyingAsset(), ethAddress)
+      assert.equal(await otoken.strikeAsset(), usdc.address)
+      assert.equal(await otoken.collateralAsset(), usdc.address)
+      assert.equal(await otoken.isPut(), isPut)
+      assert.equal((await otoken.strikePrice()).toString(), strikePrice.toString())
+      assert.equal((await otoken.expiry()).toString(), expiry.toString())
+    })
+
+    it('Should revert when creating non-whitelisted options', async () => {
+      await expectRevert(
+        otokenFactory.createOtoken(shitcoin.address, usdc.address, usdc.address, strikePrice, expiry, isPut),
+        'OtokenFactory: Unsupported Product',
+      )
+    })
+
+    it('Should revert when calling init on already inited otoken', async () => {
+      await expectRevert(otoken.init(usdc.address, usdc.address, usdc.address, strikePrice, expiry, isPut), 'revert')
+    })
+
+    it('Should revert when creating duplicated option', async () => {
+      await expectRevert(
+        otokenFactory.createOtoken(ethAddress, usdc.address, usdc.address, strikePrice, expiry, isPut),
+        'OtokenFactory: Option created',
+      )
+    })
+  })
+
+  describe('Get otoken address after creation', () => {
+    it('Should have one otoken record', async () => {
+      const otokens = await otokenFactory.getOtokens()
+      assert.equal(otokens.length, 1)
+
+      assert(otokens.includes(otoken.address))
+    })
+
+    it('should get same address if calling getTargetOTokenAddress with existing option paramters', async () => {
+      const addr = await otokenFactory.getTargetOtokenAddress(
+        ethAddress,
+        usdc.address,
+        usdc.address,
+        strikePrice,
+        expiry,
+        isPut,
+      )
+      assert.equal(addr, otoken.address)
+    })
+
+    it('Should return correct token address', async () => {
+      const existAddress = await otokenFactory.getOtoken(
+        ethAddress,
+        usdc.address,
+        usdc.address,
+        strikePrice,
+        expiry,
+        isPut,
+      )
+      assert.equal(existAddress, otoken.address)
+    })
+  })
+
+  describe('Wrong setup: wrong implementation contract', () => {
+    it('Should revert on token creation', async () => {
+      // Set the otoken Impl contract to a wrong address
+      await addressBook.setOtokenImpl(otokenFactory.address)
```

Comment:
does this test that only oTokenlogic contracts can be created?

Reply:
yeah, it tests that the creation will fail if the address book has the wrong `otokenImpl` address

Reply:
it's in `Create2.sol` 
```
require(addr != address(0), "Create2: Failed on deploy");        
```

I'm gonna update the revert message to this one